datafusion/optimizer/src/simplify_expressions/expr_simplifier.rs

```diff
@@ -734,19 +744,33 @@ impl<'a, S: SimplifyInfo> TreeNodeRewriter for Simplifier<'a, S> {
                 op: Modulo,
                 right: _,
             }) if is_null(&left) => *left,
-            // A % 1 --> 0
+            // A % 1 --> 0 (if A is not nullable and not floating, since NAN % 1 --> NAN)
             Expr::BinaryExpr(BinaryExpr {
                 left,
                 op: Modulo,
                 right,
-            }) if !info.nullable(&left)? && is_one(&right) => lit(0),
-            // A % 0 --> DivideByZero Error
+            }) if !info.nullable(&left)?
+                && !info.get_data_type(&left)?.is_floating()
+                && is_one(&right) =>
+            {
+                lit(0)
+            }
+            // A % 0 --> DivideByZero Error (if A is not floating and not null)
+            // A % 0 --> NAN (if A is floating and not null)
             Expr::BinaryExpr(BinaryExpr {
                 left,
                 op: Modulo,
                 right,
             }) if !info.nullable(&left)? && is_zero(&right) => {
-                return Err(DataFusionError::ArrowError(ArrowError::DivideByZero));
+                match info.get_data_type(&left)? {
```

Comment:
I don't really understand the rationale for `float % float` --> `NaN` (rather than error)


But on the other hand, postgres doesn't seem to support `%` on floating point values:

```
select 1.0::float % 0::float;

operator does not exist: double precision % double precision LINE 1: select 1.0::float % 0::float; ^ HINT: No operator matches the given name and argument types. You might need to add explicit type casts.

```

Reply:
@alamb 
DataFusion utilizes the `rem()` function from the arrow-rs to perform the `Modulo` operation.

The modification here ensures that it behaves the same as the `rem()` function in arrow-rs, i.e., `float % 0.` --> `NAN`.
https://github.com/apache/arrow-rs/blob/77455d48cd6609045a4728ba908123de9d0b62fd/arrow-arith/src/numeric.rs#L71-L77

And in the IEEE 754-2008 standard：
>7.2 Invalid operation 7.2.0
For operations producing results in floating-point format, the default result of an operation that signals the
invalid operation exception shall be a quiet NaN...
These operations are:
...
f) remainder: remainder(x, y), when y is zero or x is infinite...
...

Ref: https://en.wikipedia.org/wiki/NaN#Operations_generating_NaN